src/ui/CanvasTextureRenderer.ts

```diff
@@ -0,0 +1,556 @@
+  show(): void;
+  hide(): void;
+  updateSelection(dataIndices: number[]): void;
+  addRanking(ranking: EngineRanking): void;
+  removeRanking(ranking: Ranking | null): void;
+  s2d(): void;
+  d2s(): void;
+}
+
+export default class CanvasTextureRenderer implements ITextureRenderer {
+  static readonly MAX_CANVAS_SIZE = 32767;
+  static readonly AGGREGATED_ROW_HEIGHT = 45;
+  static readonly RENDER_ROW_PADDING = 10;
+  static readonly EXPANDED_ROW_CLASS = 'expand';
+  static readonly SELECTION_DRAW_WIDTH = 2;
+
+  readonly node: HTMLElement;
+  readonly canvas: any;
+  readonly headerNode: HTMLElement;
+  private renderedColumns: any[];
+  private dragStartPosition: [number, number] = [0,0];
+  private dragOverlay: HTMLElement;
+  private detailParts: any[];
+  private currentRankings: EngineRanking[] = [];
+  private currentLocalData: IDataRow[][] = [];
+  private currentNodeHeight: number = 0;
+  private currentRankingWidths: number[] = [];
+  private engineRenderer: EngineRenderer;
+  private engineRankings: EngineRanking[][] = [];
+  private alreadyExpanded: boolean = false;
+  private expandLaterRows: any[] = [];
+  private readonly options: Readonly<ILineUpOptions>;
+  private readonly idPrefix = 'testprefix';
+
+  constructor(parent: Element, engineRenderer: EngineRenderer, options: Readonly<ILineUpOptions>) {
+    this.node = parent.ownerDocument!.createElement('main');
+    this.node.id = 'lu-texture-container';
+    parent.appendChild(this.node);
+    this.canvas = parent.ownerDocument!.createElement('canvas');
+    this.headerNode = <HTMLElement>d3Select(parent).select('header').node();
+    this.engineRenderer = engineRenderer;
+    this.options = options;
+    this.renderedColumns = [];
+    this.detailParts = [];
+    this.dragOverlay = this.node;
+
+    this.node.addEventListener('scroll', () => {
+      {
+        //scroll header with main panel
+        this.headerNode.scrollLeft = this.node.scrollLeft;
+      }
+    });
+  }
+
+  updateSelection(dataIndices: number[]) {
+    const s = new Set(dataIndices);
+    this.engineRankings.forEach((v) => v.forEach((r) => r.updateSelection(s)));
+    this.drawSelection();
+    this.update();
+  }
+
+  update(rankings: EngineRanking[] = this.currentRankings, localData: IDataRow[][] = this.currentLocalData) {
+    this.detailParts = [];
+    rankings.forEach((r, i) => {
+      const rankingIndex = this.currentRankings.findIndex((v) => v === r);
+      this.currentLocalData[rankingIndex] = localData[i];
+    });
+    this.currentNodeHeight = this.node.offsetHeight;
+    let totalWidth = 0;
+    rankings.forEach((r, i) => {
+      let rankingWidth = 0;
+      r.ranking.flatColumns.forEach((c) => 'children' in c ? rankingWidth += (<CompositeColumn>c).children.length : rankingWidth += c.getWidth() + 5);
+      this.currentRankingWidths[i] = rankingWidth;
+      totalWidth += rankingWidth;
+    });
+    if (totalWidth > this.node.clientWidth) {
+        this.currentNodeHeight -= 20;
```

Comment:
Where does this 20 px are coming from? Is there an existing constant that you can use?